src/Libraries/DSCPython/DCPythonCodeCompletionProviderCore.cs

```diff
@@ -0,0 +1,394 @@
+
+            if (!PythonEngine.IsInitialized)
+            {
+                PythonEngine.Initialize();
+                PythonEngine.BeginAllowThreads();
+            }
+
+            try
+            {
+                using (Py.GIL())
+                {
+                    if (Scope == null)
+                    {
+                        Scope = CreateUniquePythonScope();
+                    }
+
+                    var assemblies = AppDomain.CurrentDomain.GetAssemblies();
+
+                    // Determine if the Revit API is available in the given context
+                    if (assemblies.Any(x => x.GetName().Name == "RevitAPI"))
+                    {
+                        // Try to load Revit Type for autocomplete
+                        try
+                        {
+                            var revitImports =
+                                "import clr\nclr.AddReference('RevitAPI')\nclr.AddReference('RevitAPIUI')\nfrom Autodesk.Revit.DB import *\nimport Autodesk\n";
+
+                            Scope.Exec(revitImports);
+
+                            ClrModules.Add("RevitAPI");
+                            ClrModules.Add("RevitAPIUI");
+                        }
+                        catch
+                        {
+                            Log("Failed to load Revit types for autocomplete. Python autocomplete will not see Autodesk namespace types.");
+                        }
+                    }
+
+                    // Determine if the ProtoGeometry is available in the given context
+                    if (assemblies.Any(x => x.GetName().Name == "ProtoGeometry"))
+                    {
+                        // Try to load ProtoGeometry Type for autocomplete
+                        try
+                        {
+                            var libGImports =
+                                "import clr\nclr.AddReference('ProtoGeometry')\nfrom Autodesk.DesignScript.Geometry import *\n";
+
+                            Scope.Exec(libGImports);
+                            ClrModules.Add("ProtoGeometry");
+                        }
+                        catch (Exception e)
+                        {
+                            Log(e.ToString());
+                            Log("Failed to load ProtoGeometry types for autocomplete. Python autocomplete will not see Autodesk namespace types.");
+                        }
+                    }
+                }
+            }
+            catch {}
+        }
+        #endregion
+
+        #region ILogSource Implementation
+        /// <summary>
+        /// Raise this event to request loggers log this message.
+        /// </summary>
+        public event Action<ILogMessage> MessageLogged;
+
+        internal void Log(string message)
+        {
+            MessageLogged?.Invoke(LogMessage.Info(message));
+        }
+
+        public void Dispose()
+        {
+            if (Scope != null)
```

Comment:
so - does `uniquePythonScopeName` ever get called more than once during the lifetime of this providerCore - if so, then shouldn't dispose somehow keep track of all the guids and dispose all the scopes - not just the last one?

Reply:
uniquePythonScopeName is assigned a unique ID only on DCPythonCodeCompletionProviderCore constructor calls and only one Scope is created during the lifetime of the provider core.
So it makes sense to only cleanup that one Scope